File: plugins/os/_llm.py
"""LLM 调用插件 — 协议 v3.0 系统插件
封装 LLM API 调用，支持 chat、chat_json。
其他插件可通过 HTTP 或 system.call 调用。
"""
import json
import aiohttp


async def execute(envelop, agent):
    action = envelop.payload.get("action", "chat")
    
    if action == "chat":
        messages = envelop.payload.get("messages", [])
        model = envelop.payload.get("model", "")
        temperature = envelop.payload.get("temperature", 0.7)
        
        if not messages:
            envelop.payload = {"error": "No messages"}
            return envelop
        
        # 从 agent.config 获取 LLM 配置
        models_config = agent.config.get("models", {})
        providers = models_config.get("providers", {})
        
        if not providers:
            envelop.payload = {"error": "No LLM providers configured"}
            return envelop
        
        # 取第一个 provider（可扩展为按 model 名匹配）
        provider_name, provider_config = next(iter(providers.items()))
        api_url = provider_config.get("api_url", "")
        api_key = provider_config.get("api_key", "")
        
        if not model:
            model = provider_config.get("default_model", "gpt-3.5-turbo")
        
        headers = {
            "Content-Type": "application/json",
            "Authorization": f"Bearer {api_key}",
        }
        
        body = {
            "model": model,
            "messages": messages,
            "temperature": temperature,
        }
        
        try:
            async with aiohttp.ClientSession() as session:
                async with session.post(api_url, headers=headers, json=body, timeout=120) as resp:
                    data = await resp.json()
                    
                    if "choices" in data:
                        envelop.payload = {
                            "content": data["choices"][0]["message"]["content"],
                            "model": model,
                            "usage": data.get("usage", {}),
                        }
                    elif "error" in data:
                        envelop.payload = {"error": data["error"]}
                    else:
                        envelop.payload = {"raw": data}
        except aiohttp.ClientError as e:
            envelop.payload = {"error": f"LLM request failed: {str(e)}"}
        except Exception as e:
            envelop.payload = {"error": f"LLM error: {str(e)}"}
        
        return envelop
    
    elif action == "chat_json":
        messages = envelop.payload.get("messages", [])
        # 注入 JSON 指令
        json_messages = [
            {"role": "system", "content": "You must respond with valid JSON only. No markdown, no explanations."},
            *messages,
        ]
        envelop.payload["messages"] = json_messages
        envelop.payload["action"] = "chat"
        
        result = await execute(envelop, agent)
        
        if "content" in result.payload:
            try:
                result.payload["json"] = json.loads(result.payload["content"])
            except json.JSONDecodeError:
                result.payload["json"] = {"error": "Invalid JSON response", "raw": result.payload["content"]}
        
        return result
    
    elif action == "models":
        models_config = agent.config.get("models", {})
        envelop.payload = {"models": models_config}
        return envelop
    
    envelop.payload = {"error": f"Unknown action: {action}"}
    return envelop

File: www/aicp-core/plugins/llm.js
// plugins/common.js — 通用能力合集（仅 LLM 能力）
(function() {
    const registry = window.__aicp_registry__;

    // LLM 对话（配置通过 agent 传入）
    registry.register('llm/chat', async (env, agent) => {
        const { messages, model, temperature } = env.payload;
        const apiKey = agent.get('apiKey');
        const baseUrl = agent.get('baseUrl') || 'https://api.openai.com/v1';
        
        if (!apiKey) {
            return { ok: false, error: 'API Key not configured' };
        }
        
        try {
            const response = await fetch(`${baseUrl}/chat/completions`, {
                method: 'POST',
                headers: { 'Content-Type': 'application/json', 'Authorization': `Bearer ${apiKey}` },
                body: JSON.stringify({ model: model || 'gpt-3.5-turbo', messages, temperature: temperature ?? 0.7 }),
            });
            
            const data = await response.json();
            console.log('[llm] response:', data);
            if (data.choices?.[0]?.message?.content) {
                return { ok: true, data: data.choices[0].message.content };
            }
            return { ok: false, error: data.error?.message || 'Empty response' };
        } catch (e) {
            return { ok: false, error: e.message };
        }
    });

    // LLM 流式对话
    registry.register('llm/chatStream', async (env, agent) => {
        const { messages, model, temperature } = env.payload;
        const apiKey = agent.get('apiKey');
        const baseUrl = agent.get('baseUrl') || 'https://api.openai.com/v1';
        
        if (!apiKey) {
            return { ok: false, error: 'API Key not configured' };
        }
        
        try {
            const response = await fetch(`${baseUrl}/chat/completions`, {
                method: 'POST',
                headers: { 'Content-Type': 'application/json', 'Authorization': `Bearer ${apiKey}` },
                body: JSON.stringify({ model: model || 'gpt-3.5-turbo', messages, temperature: temperature ?? 0.7, stream: true }),
            });
            return { ok: true, data: response.body };  // 返回 ReadableStream
        } catch (e) {
            return { ok: false, error: e.message };
        }
    });

    console.log('✅ common.js 已加载 (LLM 能力)');
})();